### PR TITLE
Add Tiling Module

### PR DESCRIPTION
**Pull request by eugene123tw**
## Tiling
* Add tiling and corresponded change to the following submodule
  * mmdetection: https://github.com/openvinotoolkit/mmdetection/pull/356
  * MPA: https://github.com/openvinotoolkit/model_preparation_algorithm/pull/40

### CODE REVIEW

**Comment by harimkang** [comment]
run ote-test

**Comment by harimkang** [comment]
run ote-test

**Comment by samet-akcay** [comment]
By the way, it's PR 1200! Congrats!!! 🎉🎉🎉

**Comment by eugene123tw** [comment]
run ote-test

**Comment by eugene123tw** [comment]
run ote-test

File: external/model-preparation-algorithm/mpa_tasks/apis/task.py [hunk 1]
```diff
@@ -262,11 +259,43 @@ def _init_test_data_cfg(self, dataset: DatasetEntity) -> Union[Config, None]:
         """
         return None
 
-    def _init_recipe_hparam(self) -> dict:
-        """
-        initialize recipe hyperparamter as dict.
+    def _overwrite_parameters(self):
```
Comment (harimkang) [inline external/model-preparation-algorithm/mpa_tasks/apis/task.py:262]: Is this a replacement function for "_init_recipe_hparam"?
It seems to still remain in the classification task.

File: external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py [hunk 2]
```diff
@@ -193,40 +197,36 @@ def export(self, export_type: ExportType, output_model: ModelEntity):
...
+                )
             )
-        else:
-            early_stop = False
-
-        return ConfigDict(
-            optimizer=ConfigDict(lr=self._hyperparams.learning_parameters.learning_rate),
-            lr_config=lr_config,
-            early_stop=early_stop,
-            use_adaptive_interval=self._hyperparams.learning_parameters.use_adaptive_interval,
```
Comment (harimkang) [inline external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py:223]: This is a setting (use_adaptive_interval) that exists only for the current detection, but will it be maintained?

File: external/model-preparation-algorithm/tests/ote_cli/test_small_object_inst_seg.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
+"""Tests for MPA Class-Incremental Learning for instance segmentation with OTE CLI"""
```
Comment (harimkang) [inline external/model-preparation-algorithm/tests/ote_cli/test_small_object_inst_seg.py]: The name 'test_small_*' doesn't come to mind at once. Any other names to replace?

**Review by harimkang — CHANGES_REQUESTED** [review]
Added some comments. 
+ It seems that this PR is not pre-commit checked.

You can check pre-commit using the script below.
tests/run_code_checks.sh

File: external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py (hunk 2 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py:223]: @harimkang Thanks Harim, I missed it. It will be maintained: 

```
self._recipe_cfg.merge_from_dict(
    dict(use_adaptive_interval=self._hyperparams.learning_parameters.use_adaptive_interval))
```

File: external/model-preparation-algorithm/mpa_tasks/apis/task.py (hunk 1 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/mpa_tasks/apis/task.py:262]: yes, it's a replacement. Instead of having `_init_recipe_hparam` in each task, it's maintained by BaseTask now and inherited and overwritten per task if desired.

File: external/model-preparation-algorithm/tests/ote_cli/test_small_object_inst_seg.py (hunk 3 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/tests/ote_cli/test_small_object_inst_seg.py]: test_tiling_inst_seg?

Comment (harimkang) [reply external/model-preparation-algorithm/tests/ote_cli/test_small_object_inst_seg.py]: I think it looks better.

**Comment by harimkang** [comment]
@jaegukhyun @JihwanEom Please review this PR.

**Review by harimkang — DISMISSED** [review]
Currently, OTX CI does not work, but it seems that you can check the minimum test locally.
```
pytest -s -v ote_sdk/ote_sdk/tests
pytest -s -v tests/ote_cli/external/model-preparation-algorithm/ote_cli
pytest -s -v external/model-preparation-algorithm/tests/api_tests
```
These are ote_sdk unit test, mpa api test, and ote cli test in order from the top.
(These are the simplest tests that exclude training tests that require a separate dataset)

**Comment by eugene123tw** [comment]
run ote-test

**Comment by harimkang** [comment]
run ote-test

File: external/model-preparation-algorithm/configs/detection/configuration.yaml [hunk 4]
```diff
@@ -390,3 +390,96 @@ nncf_optimization:
...
+    visible_in_ui: true
+    warning: null
+
+  tile_size:
+    header: Tile Image Size
+    description: Tile Image Size
+    affects_outcome_of: TRAINING
+    default_value: 400
+    min_value: 100
+    max_value: 9223372036854775807
```
Comment (JihwanEom) [inline external/model-preparation-algorithm/configs/detection/configuration.yaml]: Minor question :) what was the criteria to set max value?

File: external/model-preparation-algorithm/configs/detection/mobilenetv2_atss_cls_incr/tile_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+dataset_type = "CocoDataset"
+
+img_size = (992, 736)
+
+tile_cfg = dict(
+    tile_size=400, min_area_ratio=0.9, overlap_ratio=0.2, iou_threshold=0.45, max_per_img=1500, filter_empty_gt=True
+)
+
+img_norm_cfg = dict(mean=[0, 0, 0], std=[255, 255, 255], to_rgb=True)
```
Comment (JihwanEom) [inline external/model-preparation-algorithm/configs/detection/mobilenetv2_atss_cls_incr/tile_pipeline.py:9]: Why image normalize config is different with yolox? 
`img_norm_cfg = dict(mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)`

File: external/model-preparation-algorithm/configs/detection/mobilenetv2_ssd_cls_incr/tile_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
+dataset_type = "CocoDataset"
+
+img_size = (864, 864)
+
+tile_cfg = dict(
+    tile_size=400, min_area_ratio=0.9, overlap_ratio=0.2, iou_threshold=0.45, max_per_img=1500, filter_empty_gt=True
+)
+
+img_norm_cfg = dict(mean=[0, 0, 0], std=[255, 255, 255], to_rgb=True)
```
Comment (JihwanEom) [inline external/model-preparation-algorithm/configs/detection/mobilenetv2_ssd_cls_incr/tile_pipeline.py:9]: Same with here

File: external/model-preparation-algorithm/configs/detection/resnet50_vfnet_cls_incr/tile_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
+dataset_type = "CocoDataset"
+
+img_size = (1024, 1024)
+
+tile_cfg = dict(
+    tile_size=400, min_area_ratio=0.9, overlap_ratio=0.2, iou_threshold=0.45, max_per_img=1500, filter_empty_gt=True
+)
+
+img_norm_cfg = dict(mean=(103.53, 116.28, 123.675), std=(1.0, 1.0, 1.0), to_rgb=False)
```
Comment (JihwanEom) [inline external/model-preparation-algorithm/configs/detection/resnet50_vfnet_cls_incr/tile_pipeline.py]: Could you align normalize vactor and rgb/bgr?

File: external/model-preparation-algorithm/configs/detection/mobilenetv2_atss_cls_incr/tile_pipeline.py (hunk 5 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/configs/detection/mobilenetv2_atss_cls_incr/tile_pipeline.py:9]: These are the default setting for ATSS and YOLOX. But I'm guessing these were the normalise values that were used in training the pre-trained COCO model.

File: external/model-preparation-algorithm/configs/detection/mobilenetv2_ssd_cls_incr/tile_pipeline.py (hunk 6 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/configs/detection/mobilenetv2_ssd_cls_incr/tile_pipeline.py:9]: It's the default setting: https://github.com/openvinotoolkit/training_extensions/blob/develop/external/model-preparation-algorithm/configs/detection/mobilenetv2_ssd_cls_incr/data_pipeline.py#L3 and the same reason mentioned above

File: external/model-preparation-algorithm/configs/detection/resnet50_vfnet_cls_incr/tile_pipeline.py (hunk 7 above)
Comment (eugene123tw) [reply external/model-preparation-algorithm/configs/detection/resnet50_vfnet_cls_incr/tile_pipeline.py]: Done

**Review by harimkang — APPROVED** [review]

**Review by JihwanEom — APPROVED** [review]
